include/swift/Runtime/Metadata.h

```diff
@@ -2867,6 +2868,16 @@ void swift_initStructMetadata_UniversalStrategy(size_t numFields,
                                          size_t *fieldOffsets,
                                          ValueWitnessTable *vwtable);
 
+/// Relocate the metadata for a class and copy fields from the given template.
+/// The final size of the metadata is calculated at runtime from the size of
+/// the superclass metadata together with the given number of immediate
+/// members.
+SWIFT_RUNTIME_EXPORT
+ClassMetadata *
+swift_relocateClassMetadata(ClassMetadata *self,
+                            size_t templateSize,
+                            size_t numImmediateMembers);
+
```

Comment:
If class relocation ever requires class-specific treatment, rather than just memcpy'ing a certain number of bytes, how do we determine the class corresponding to the template?  Is it always the most-derived class?

Reply:
And if we can reliably figure that out from the information provided, why exactly do numImmediateMembers and templateSize need to be passed in?

Reply:
It's always the most-derived class, yeah. They're passed in because they're not stored in the class metadata anywhere in the case where the class is not generic but resilient for some other reason (inherits a generic class, or a resilient class from another module).

Note that in this case and in the generic case, the "template" has very little in it, basically the metadata header and some of the Objective-C stuff. I think it would be good to get rid of the memcpy() altogether, and use a more efficient representation of the template that uses relative pointers and omits any fields filled in at runtime entirely.

However I'm hoping that @jckarter's work on cleaning up type metadata can lead to this soon, and for now I kind of want to prototype the right behavior for class resilience, even if the ABI is not final.

Reply:
Yes, the idea of not doing a blind memcpy at all has some appeal.

Reply:
Also, yeah, it seems to me that we're playing a very weird game with headers and templates, where the header really has to treat the template as opaque data but has fields that suggest that it doesn't.

Reply:
Yep. I'd like to circle back to this soon.